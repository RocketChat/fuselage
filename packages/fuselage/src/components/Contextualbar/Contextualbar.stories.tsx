import { action } from '@storybook/addon-actions';
import type { ComponentStory, ComponentMeta } from '@storybook/react';
import React from 'react';

import {
  Contextualbar,
  ContextualbarAction,
  ContextualbarActions,
  ContextualbarButton,
  ContextualbarContent,
  ContextualbarEmptyContent,
  ContextualbarFooter,
  ContextualbarHeader,
  ContextualbarSkeleton,
  ContextualbarTitle,
} from '.';
import { Button, ButtonGroup, IconButton, Box } from '..';

export default {
  title: 'Containers/Contextualbar',
  component: Contextualbar,
  parameters: {
    docs: {
      description: {
        component: `The \`Contextualbar\` has the purpose to persist and input information about the scope of the related page.
				`,
      },
    },
  },
  decorators: [
    (storyFn) => (
      <Box width='x400' elevation='2'>
        {storyFn()}
      </Box>
    ),
  ],
} as ComponentMeta<typeof Contextualbar>;

export const Default: ComponentStory<typeof Contextualbar> = () => (
  <Contextualbar position='static' height='x540'>
    <ContextualbarHeader>
      <ContextualbarAction
        title='Back'
        name='arrow-back'
        onClick={action('click')}
      />
      <ContextualbarTitle>Contextualbar Title</ContextualbarTitle>
      <ContextualbarActions>
        <ContextualbarAction
          title='Title'
          name='new-window'
          onClick={action('click')}
        />
        <ContextualbarAction
          title='Add user'
          name='add-user'
          onClick={action('click')}
        />
      </ContextualbarActions>
    </ContextualbarHeader>
    <ContextualbarContent>Contextualbar Content</ContextualbarContent>
    <ContextualbarFooter>
      <ButtonGroup>
        <ContextualbarButton width='full' secondary>
          Cancel
        </ContextualbarButton>
        <Button width='full' primary>
          Save
        </Button>
        <IconButton title='More' icon='menu' />
      </ButtonGroup>
    </ContextualbarFooter>
  </Contextualbar>
);

export const Skeleton: ComponentStory<typeof Contextualbar> = () => (
  <ContextualbarSkeleton position='static' height='x540' />
);

export const Empty: ComponentStory<typeof Contextualbar> = () => (
  <Contextualbar position='static' height='x540'>
    <ContextualbarHeader>
      <ContextualbarAction title='Back' name='arrow-back' />
      <ContextualbarTitle>Contextualbar Empty</ContextualbarTitle>
      <ContextualbarActions>
        <ContextualbarAction
          title='Title'
          name='new-window'
          onClick={action('click')}
        />
      </ContextualbarActions>
    </ContextualbarHeader>
    <ContextualbarEmptyContent />
    <ContextualbarFooter>Footer</ContextualbarFooter>
  </Contextualbar>
);
