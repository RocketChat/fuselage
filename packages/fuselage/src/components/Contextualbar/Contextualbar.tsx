import type { ComponentProps } from 'react';
import React, { memo } from 'react';

import { Box } from '..';

type ContextualbarProps = ComponentProps<typeof Box>;

const Contextualbar = ({
  children,
  width,
  position,
  bg = 'room',
  ...props
}: ContextualbarProps) => (
  <Box
    rcx-vertical-bar
    bg={bg}
    color='default'
    display='flex'
    flexDirection='column'
    flexShrink={0}
    width={width}
    borderInlineStartWidth='default'
    borderInlineStartColor='extra-light'
    borderInlineStartStyle='solid'
    height='full'
    position={position}
    insetInlineEnd='none'
    insetBlockStart='none'
    zIndex={5}
    {...props}
  >
    {children}
  </Box>
);

export default memo(Contextualbar);
