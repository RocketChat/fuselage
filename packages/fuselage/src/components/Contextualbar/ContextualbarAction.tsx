import type { ReactElement, ComponentProps } from 'react';
import React, { memo } from 'react';

import type { Icon } from '..';
import { IconButton } from '..';

type ContextualbarActionProps = {
  name: ComponentProps<typeof Icon>['name'];
} & Omit<ComponentProps<typeof IconButton>, 'icon'>;

const ContextualbarAction = ({
  name,
  ...props
}: ContextualbarActionProps): ReactElement => (
  <IconButton {...props} small flexShrink={0} icon={name} />
);

export default memo(ContextualbarAction);
