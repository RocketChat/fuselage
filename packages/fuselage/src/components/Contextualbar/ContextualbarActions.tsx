import type { ReactElement, ComponentProps } from 'react';
import React, { memo } from 'react';

import { ButtonGroup } from '..';

const ContextualbarActions = (
  props: ComponentProps<typeof ButtonGroup>
): ReactElement => <ButtonGroup {...props} />;

export default memo(ContextualbarActions);
