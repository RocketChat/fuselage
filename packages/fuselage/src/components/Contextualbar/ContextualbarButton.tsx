import type { ComponentProps, ReactElement } from 'react';
import React, { memo } from 'react';

import { Button } from '..';

const ContextualbarButton = (
  props: ComponentProps<typeof Button>
): ReactElement => <Button {...props} />;

export default memo(ContextualbarButton);
