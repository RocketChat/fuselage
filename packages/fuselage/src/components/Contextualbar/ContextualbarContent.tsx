import type { ForwardedRef } from 'react';
import React, { forwardRef, memo } from 'react';

import { Box, type BoxProps } from '../Box';

type ContextualbarContentProps = BoxProps;

const ContextualbarContent = forwardRef(function ContextualbarContent(
  props: ContextualbarContentProps,
  ref: ForwardedRef<HTMLElement>
) {
  return (
    <Box
      ref={ref}
      rcx-vertical-bar__content
      paddingInline={24}
      display='flex'
      flexDirection='column'
      overflowY='hidden'
      height='full'
      {...props}
    />
  );
});

export default memo(ContextualbarContent);
