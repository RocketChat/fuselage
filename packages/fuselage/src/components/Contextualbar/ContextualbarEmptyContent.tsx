import type { IconName } from '@rocket.chat/icons';
import React, { forwardRef, memo } from 'react';

import type { BoxProps } from '../Box';
import { States, StatesIcon, StatesSubtitle, StatesTitle } from '../States';
import ContextualbarContent from './ContextualbarContent';

type ContextualbarEmptyContentProps = BoxProps & {
  icon?: IconName;
  title?: string;
  subtitle?: string;
};

const ContextualbarEmptyContent = forwardRef<
  HTMLElement,
  ContextualbarEmptyContentProps
>(function ContextualbarEmptyContent(
  { icon = 'magnifier', title = 'Nothing Found', subtitle, ...props },
  ref
) {
  return (
    <ContextualbarContent justifyContent='center' {...props} ref={ref}>
      <States>
        <StatesIcon name={icon} />
        <StatesTitle>{title}</StatesTitle>
        {subtitle && <StatesSubtitle>{subtitle}</StatesSubtitle>}
      </States>
    </ContextualbarContent>
  );
});

export default memo(ContextualbarEmptyContent);
