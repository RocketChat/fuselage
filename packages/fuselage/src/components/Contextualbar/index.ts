import Contextualbar from './Contextualbar';
import ContextualbarAction from './ContextualbarAction';
import ContextualbarActions from './ContextualbarActions';
import ContextualbarButton from './ContextualbarButton';
import ContextualbarContent from './ContextualbarContent';
import ContextualbarEmptyContent from './ContextualbarEmptyContent';
import ContextualbarFooter from './ContextualbarFooter';
import ContextualbarHeader from './ContextualbarHeader';
import ContextualbarIcon from './ContextualbarIcon';
import ContextualbarSection from './ContextualbarSection';
import ContextualbarSkeleton from './ContextualbarSkeleton';
import ContextualbarTitle from './ContextualbarTitle';

export {
  Contextualbar,
  ContextualbarAction,
  ContextualbarActions,
  ContextualbarButton,
  ContextualbarContent,
  ContextualbarEmptyContent,
  ContextualbarFooter,
  ContextualbarHeader,
  ContextualbarIcon,
  ContextualbarSkeleton,
  ContextualbarTitle,
  ContextualbarSection,
};
