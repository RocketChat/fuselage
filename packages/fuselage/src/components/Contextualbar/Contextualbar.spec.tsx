import { composeStories } from '@storybook/testing-react';
import { render } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import React from 'react';

import * as stories from './Contextualbar.stories';

expect.extend(toHaveNoViolations);

const testCases = Object.values(composeStories(stories)).map((Story) => [
  Story.storyName || 'Story',
  Story,
]);

describe('[Contextualbar Rendering]', () => {
  test.each(testCases)(
    `renders %s without crashing`,
    async (_storyname, Story) => {
      const tree = render(<Story />);
      expect(tree.baseElement).toMatchSnapshot();
    }
  );

  test.each(testCases)(
    '%s should have no a11y violations',
    async (_storyname, Story) => {
      const { container } = render(<Story />);

      const results = await axe(container);
      expect(results).toHaveNoViolations();
    }
  );
});
