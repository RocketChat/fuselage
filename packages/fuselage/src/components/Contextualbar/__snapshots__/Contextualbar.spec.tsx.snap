// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Contextualbar Rendering] renders Default without crashing 1`] = `
<body>
  <div>
    <div
      class="rcx-box rcx-box--full rcx-css-fs953a"
    >
      <div
        class="rcx-box rcx-box--full rcx-vertical-bar rcx-css-1pzjc7w"
      >
        <div
          class="rcx-box rcx-box--full rcx-css-13kuqi8"
        >
          <div
            class="rcx-box rcx-box--full rcx-css-1sl6k6j"
          >
            <button
              class="rcx-box rcx-box--full rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-css-x7bl3q rcx-css-1yzvz7u"
              title="Back"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-icon--name-arrow-back rcx-icon rcx-css-4pvxx3"
              >
                ⦉
              </i>
            </button>
            <div
              class="rcx-box rcx-box--full  rcx-css-x7bl3q rcx-css-1to6ka7"
            >
              Contextualbar Title
            </div>
            <div
              class="rcx-button-group rcx-button-group--align-start rcx-css-x7bl3q"
              role="group"
            >
              <button
                class="rcx-box rcx-box--full rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item rcx-css-1yzvz7u"
                title="Title"
                type="button"
              >
                <i
                  aria-hidden="true"
                  class="rcx-box rcx-box--full rcx-icon--name-new-window rcx-icon rcx-css-4pvxx3"
                >
                  ⁽
                </i>
              </button>
              <button
                class="rcx-box rcx-box--full rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item rcx-css-1yzvz7u"
                title="Add user"
                type="button"
              >
                <i
                  aria-hidden="true"
                  class="rcx-box rcx-box--full rcx-icon--name-add-user rcx-icon rcx-css-4pvxx3"
                >
                  
                </i>
              </button>
            </div>
          </div>
        </div>
        <div
          class="rcx-box rcx-box--full rcx-vertical-bar__content rcx-css-1ptgac4"
        >
          Contextualbar Content
        </div>
        <div
          class="rcx-box rcx-box--full rcx-css-m843eh"
        >
          <div
            class="rcx-button-group rcx-button-group--align-start"
            role="group"
          >
            <button
              class="rcx-box rcx-box--full rcx-button--secondary rcx-button  rcx-button-group__item rcx-css-1qcz93u"
              type="button"
            >
              <span
                class="rcx-button--content"
              >
                Cancel
              </span>
            </button>
            <button
              class="rcx-box rcx-box--full rcx-button--primary rcx-button  rcx-button-group__item rcx-css-1qcz93u"
              type="button"
            >
              <span
                class="rcx-button--content"
              >
                Save
              </span>
            </button>
            <button
              class="rcx-box rcx-box--full rcx-button--large-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item"
              title="More"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-icon--name-menu rcx-icon rcx-css-1x2l7ts"
              >
                
              </i>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`[Contextualbar Rendering] renders Empty without crashing 1`] = `
<body>
  <div>
    <div
      class="rcx-box rcx-box--full rcx-css-fs953a"
    >
      <div
        class="rcx-box rcx-box--full rcx-vertical-bar rcx-css-1pzjc7w"
      >
        <div
          class="rcx-box rcx-box--full rcx-css-13kuqi8"
        >
          <div
            class="rcx-box rcx-box--full rcx-css-1sl6k6j"
          >
            <button
              class="rcx-box rcx-box--full rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-css-x7bl3q rcx-css-1yzvz7u"
              title="Back"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-icon--name-arrow-back rcx-icon rcx-css-4pvxx3"
              >
                ⦉
              </i>
            </button>
            <div
              class="rcx-box rcx-box--full  rcx-css-x7bl3q rcx-css-1to6ka7"
            >
              Contextualbar Empty
            </div>
            <div
              class="rcx-button-group rcx-button-group--align-start rcx-css-x7bl3q"
              role="group"
            >
              <button
                class="rcx-box rcx-box--full rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item rcx-css-1yzvz7u"
                title="Title"
                type="button"
              >
                <i
                  aria-hidden="true"
                  class="rcx-box rcx-box--full rcx-icon--name-new-window rcx-icon rcx-css-4pvxx3"
                >
                  ⁽
                </i>
              </button>
            </div>
          </div>
        </div>
        <div
          class="rcx-box rcx-box--full rcx-vertical-bar__content rcx-css-r8c38r"
        >
          <div
            class="rcx-states"
          >
            <i
              aria-hidden="true"
              class="rcx-box rcx-box--full rcx-states__icon rcx-icon--name-magnifier rcx-icon rcx-css-d884y5"
            >
              
            </i>
            <h3
              class="rcx-states__title"
            >
              Nothing Found
            </h3>
          </div>
        </div>
        <div
          class="rcx-box rcx-box--full rcx-css-m843eh"
        >
          Footer
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`[Contextualbar Rendering] renders Skeleton without crashing 1`] = `
<body>
  <div>
    <div
      class="rcx-box rcx-box--full rcx-css-fs953a"
    >
      <div
        class="rcx-box rcx-box--full rcx-vertical-bar rcx-css-1pzjc7w"
      >
        <div
          class="rcx-box rcx-box--full rcx-css-13kuqi8"
        >
          <div
            class="rcx-box rcx-box--full rcx-css-1sl6k6j"
          >
            <span
              class="rcx-box rcx-box--full rcx-skeleton--text rcx-skeleton  rcx-css-x7bl3q rcx-css-16xwxf4"
            />
          </div>
        </div>
        <div
          class="rcx-box rcx-box--full rcx-css-m843eh"
        >
          <span
            class="rcx-box rcx-box--full rcx-skeleton--rect rcx-skeleton rcx-css-kq1jj5"
          />
          <span
            class="rcx-box rcx-box--full rcx-skeleton--text rcx-skeleton"
          />
          <span
            class="rcx-box rcx-box--full rcx-skeleton--text rcx-skeleton"
          />
          <span
            class="rcx-box rcx-box--full rcx-skeleton--text rcx-skeleton"
          />
          <span
            class="rcx-box rcx-box--full rcx-skeleton--text rcx-skeleton"
          />
          <span
            class="rcx-box rcx-box--full rcx-skeleton--text rcx-skeleton"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
